From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 9 Dec 2010 20:54:07 -0500
Subject: [scsi] qla4xxx: no device add until scsi_add_host success
Message-id: <20101209205406.10052.79947.sendpatchset@localhost.localdomain>
Patchwork-id: 30058
O-Subject: [RHEL 5.6 PATCH 1/3] qla4xxx: Do not add device dynamically untill
	scsi_add_host is successful.
Bugzilla: 661768

Bugzilla
--------

Bug 661768 (https://bugzilla.redhat.com/show_bug.cgi?id=661768)

Upstream Status
---------------

RHEL 5 specific as the RHEL 5 driver uses a different mechanism for scanning for targets.

Description
-----------

This fixes the panic:

Unable to handle kernel NULL pointer dereference at 0000000000000010 RIP:
 [<ffffffff8011070a>] create_dir+0x11/0x1cf
PGD 7ef1d067 PUD 7ef3c067 PMD 0
Oops: 0000 [1] SMP
last sysfs file: /block/ram0/dev
CPU 2
Modules linked in: qla4xxx(U) scsi_transport_iscsi2 scsi_transport_iscsi shpchp
mptspi mptscsih mptbase scsi_transport_spi sd_mod scsi_mod ext3 jbd uhci_hcd ohci_hcd ehci_hcd
Pid: 498, comm: insmod Tainted: G      2.6.18-231.el5 #1
RIP: 0010:[<ffffffff8011070a>]  [<ffffffff8011070a>] create_dir+0x11/0x1cf12>
RSP: 0000:ffff81007efc3af8  EFLAGS: 00010282
RAX: ffff81007e870268 RBX: ffff81007efe8b10 RCX: ffff81007efc3b30
RDX: ffff81007efe8b18 RSI: 0000000000000000 RDI: ffff81007efe8b10
RBP: ffff81007efe8b10 R08: 00000000ffffffff R09: 0000000000000200
R10: 00000000ffffffff R11: 0000000000000000 R
MPTBIOS-5.06.06
Call Trace:

HBA ID
 [<ffffffff80110c93>] sysfs_create_dir+0x58/0x76E
 [<ffffffff80154100>] kobject_add+0xd2/0x191------ ----------------
-------- ---  [<ffffffff8813cfd8>]
:qla4xxx:qla4xxx_process_ddb_changed+0x10a/0x23c
 [<ffffffff8813fee0>] :qla4xxx:qla4xxx_process_aen+0x1f0/0x220
 [<ffffffff8813ccc8>] :qla4xxx:qla4xxx_initialize_ddb_list+0x474/0x48e
 [<ffffffff8813c6b4>] :qla4xxx:qla4xxx_init_firmware+0x726/0x755
 [<ffffffff8813ce0c>] :qla4xxx:qla4xxx_initialize_adapter+0x12a/0x1ec
 [<ffffffff88139db6>] :qla4xxx:qla4xxx_probe_adapter+0x42d/0x80f
 [<ffffffff801614e8>] pci_device_probe+0x104/0x184
[<ffffffff801cb828>] driver_probe_device+0x52/0xaa
[<ffffffff801cb957>] __driver_attach+0x65/0xb6  [<ffffffff801cb8f2>]
__driver_attach+0x0/0xb6  [<ffffffff801cb12f>]
bus_for_each_dev+0x43/0x6e  [<ffffffff801cad6b>]
bus_add_driver+0x76/0x110  [<ffffffff80161804>]
__pci_register_driver+0x51/0xa6  [<ffffffff8815f0cf>]
:qla4xxx:qla4xxx_module_init+0xcf/0x101
 [<ffffffff800a8d1e>] sys_init_module+0xaf/0x1f2  [<ffffffff8005d116>]
system_call+0x7e/0x83

Code: 48 8b 7e 10 48 89 d3 48 81 c7 b8 00 00 00 e8 bb 33 f5 ff fc RIP
[<ffffffff8011070a>] create_dir+0x11/0x1cf  RSP <ffff81007efc3af8>
CR2: 0000000000000010
 <0>Kernel panic - not syncing: Fatal exception

>From 721cb897961878c9921a3744beb8c9ba39a4f5ac Mon Sep 17 00:00:00 2001
From: Prasanna Mumbai <prasanna.mumbai@qlogic.com>
Date: Tue, 7 Dec 2010 19:57:48 -0800
Subject: [PATCH 1/3] qla4xxx: [Update] Do not add device dynamically untill scsi_add_host is successful.

Signed-off-by: Prasanna Mumbai <prasanna.mumbai@qlogic.com>

diff --git a/drivers/scsi/qla4xxx/ql4_init.c b/drivers/scsi/qla4xxx/ql4_init.c
index 180eb01..2b83568 100644
--- a/drivers/scsi/qla4xxx/ql4_init.c
+++ b/drivers/scsi/qla4xxx/ql4_init.c
@@ -1533,6 +1533,12 @@ int qla4xxx_process_ddb_changed(struct scsi_qla_host *ha, uint32_t fw_ddb_index,
 	}
 
 	ddb_entry->fw_ddb_device_state = state;
+
+	/* If driver probe has not yet completed, do not process
+	 * devices that have come back online or have gone away */
+	if (!test_bit(AF_PROBE_DONE, &ha->flags))
+		return QLA_SUCCESS;
+
 	/* Device is back online. */
 	if (ddb_entry->fw_ddb_device_state == DDB_DS_SESSION_ACTIVE) {
 		atomic_set(&ddb_entry->port_down_timer,
diff --git a/drivers/scsi/qla4xxx/ql4_os.c b/drivers/scsi/qla4xxx/ql4_os.c
index b73348d..ba798a0 100644
--- a/drivers/scsi/qla4xxx/ql4_os.c
+++ b/drivers/scsi/qla4xxx/ql4_os.c
@@ -2008,6 +2008,8 @@ static int __devinit qla4xxx_probe_adapter(struct pci_dev *pdev,
 				goto remove_host;
 	}
 
+	set_bit(AF_PROBE_DONE, &ha->flags);
+
 	printk(KERN_INFO
 	       " QLogic iSCSI HBA Driver version: %s\n"
 	       "  QLogic ISP%04x @ %s, pdev = %p host#=%ld, fw=%02d.%02d.%02d.%02d\n",
